pkg/kinflate/resource/util.go

```diff
@@ -90,3 +90,71 @@ func Merge(rcs ...ResourceCollection) (ResourceCollection, error) {
 
 	return all, nil
 }
+
+// Merge will merge all of the entries in the slice of ResourceCollection with Override
+// If there is already an entry with the same GVKN exists, different actions are performed according to value of Behavior field
+// 'create': create a new one;
+// 'replace': replace the data only; keep the labels and annotations
+// 'merge': merge the data; keep the labels and annotations
+func MergeWithOverride(rcs ...ResourceCollection) (ResourceCollection, error) {
+	all := ResourceCollection{}
+
+	for _, rc := range rcs {
+		for gvkn, obj := range rc {
+			if _, found := all[gvkn]; found {
+				switch obj.Behavior {
+				case "", "create":
+					return nil, fmt.Errorf("Create an existing gvkn %#v is not allowed", gvkn)
+				case "replace":
+					originalObj := all[gvkn]
+					obj.Data.SetLabels(mergeMap(originalObj.Data.GetLabels(), obj.Data.GetLabels()))
+					obj.Data.SetAnnotations(mergeMap(originalObj.Data.GetAnnotations(), obj.Data.GetAnnotations()))
+					obj.Data.SetName(originalObj.Data.GetName())
+					all[gvkn] = obj
+				case "merge":
+					originalObj := all[gvkn]
+					obj.Data.SetLabels(mergeMap(originalObj.Data.GetLabels(), obj.Data.GetLabels()))
+					obj.Data.SetAnnotations(mergeMap(originalObj.Data.GetAnnotations(), obj.Data.GetAnnotations()))
+					obj.Data.SetName(originalObj.Data.GetName())
+					mergeData(obj.Data.Object, originalObj.Data.Object, obj.Data.Object)
+					all[gvkn] = obj
+				default:
+					return nil, fmt.Errorf("The behavior of %#v must be one of create, merge and replace", gvkn)
+				}
+			} else {
+				switch obj.Behavior {
+				case "", "create", "replace":
+					all[gvkn] = obj
+				case "merge":
+					return nil, fmt.Errorf("No merge is allowed for non existing gvkn %#v", gvkn)
+				default:
+					return nil, fmt.Errorf("The behavior of %#v must be one of create, merge and replace", gvkn)
+				}
+			}
+		}
+	}
+	return all, nil
+}
+
+func mergeMap(maps ...map[string]string) map[string]string {
+	mergedMap := map[string]string{}
+	for _, m := range maps {
+		for key, value := range m {
+			mergedMap[key] = value
+		}
+	}
+	return mergedMap
+}
+
+func mergeData(mergedTo map[string]interface{}, maps ...map[string]interface{}) {
```

Comment:
I'd like to call it `mergeConfigMap` which is more descriptive.
And after we sync to new k8s.io/api, we can hangle [BinaryData](https://github.com/kubernetes/kubernetes/blob/3538676841c33c87b8790947b898b2c61bc49c9d/staging/src/k8s.io/api/core/v1/types.go#L5018-L5026) here.

And please leave a TODO somewhere about handling `BinaryData` to ensure we will not forget it. :)

Reply:
Sounds good, will add it.